pkg/metrics/metrics.go

```diff
@@ -0,0 +1,218 @@
+// Copyright Jetstack Ltd. See LICENSE for details.
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog"
+)
+
+const (
+	promNamespace = "kube_oidc_proxy"
+)
+
+type Metrics struct {
+	*http.Server
+
+	registry *prometheus.Registry
+
+	// Metrics for incoming client requests
+	clientRequests *prometheus.CounterVec
+	clientDuration *prometheus.HistogramVec
+
+	// Metrics for outgoing server requests
+	serverRequests *prometheus.CounterVec
+	serverDuration *prometheus.HistogramVec
+
+	// Metrics for authentication of incoming requests
+	oidcAuthCount *prometheus.CounterVec
+
+	// Metrics for token reviews
+	tokenReviewDuration *prometheus.HistogramVec
+}
+
+func New() *Metrics {
+	var (
+		clientRequests = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: promNamespace,
+				Name:      "http_client_requests",
+				Help:      "The number of requests for incoming requests.",
+			},
+			[]string{"code", "path", "remote_address"},
+		)
+		clientDuration = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: promNamespace,
+				Name:      "http_client_duration_seconds",
+				Help:      "The duration in seconds for incoming client requests to be responded to.",
+				Buckets:   prometheus.LinearBuckets(.000, .05, 30),
```

Comment:
I'm not sure this wants to start at `.000` as the upper bound? That seems like a wasted bucket.

Also is linear the right choice here? These durations will include a the full auth and k8s apiserver roundtrips? If so then I could imagine that they will get pretty long. I would lean towards something like 10 exponential buckets rather than 30 linear ones for that.